File: solution/0000-0099/0028.Find the Index of the First Occurrence in a String/Solution.go
func strStr(haystack string, needle string) int {
    lenP, lenQ := len(haystack), len(needle)
    if lenQ == 0 {
        return 0
    }
    if lenQ > lenP {
        return -1
    }
    p, q := 0, 0
    for p < lenP-lenQ+1 {
        if (haystack[p+q] == needle[q]) {
            if q == lenQ-1 {
                return p
            }
            q++
            continue
        }
        p++
        q = 0
    }
    return -1
}

File: solution/0000-0099/0028.Find the Index of the First Occurrence in a String/Solution.java
class Solution {
    public int strStr(String haystack, String needle) {
        int lenP = haystack.length();
        int lenQ = needle.length();
        if (lenQ == 0) {
            return 0;
        }
        if (lenQ > lenP) {
            return -1;
        }
        int p = 0;
        int q = 0;
        while (p < lenP - lenQ + 1) {
            if (haystack.charAt(p + q) == needle.charAt(q)) {
                if (++q == lenQ) {
                    return p;
                }
                continue;
            }
            p++;
            q = 0;
        }
        return -1;
    }
}
